File: project-c-cosminpac04/motorcycleApp/Repositories/ParticipantRepository.cs
// using System;
// using System.Collections.Generic;
// using System.Data.SQLite;
// using motorcycleApp.Models;
//
// namespace motorcycleApp.Repositories
// {
//     public class ParticipantRepository : IRepository<Participant, int>
//     {
//         public void Add(Participant participant)
//         {
//             using (var connection = AppDbContext.GetConnection())
//             {
//                 using (var command = new SQLiteCommand("INSERT INTO Participant (Name, EngineCapacity, Team) VALUES (@Name, @EngineCapacity, @Team)", connection))
//                 {
//                     command.Parameters.AddWithValue("@Name", participant.Name);
//                     command.Parameters.AddWithValue("@EngineCapacity", participant.EngineCapacity);
//                     command.Parameters.AddWithValue("@Team", participant.Team);
//                     command.ExecuteNonQuery();
//                 }
//             }
//         }
//
//         public void Delete(Participant participant)
//         {
//             using (var connection = AppDbContext.GetConnection())
//             {
//                 using (var command = new SQLiteCommand("DELETE FROM Participant WHERE ID = @ID", connection))
//                 {
//                     command.Parameters.AddWithValue("@ID", participant.ID); // Use the participant's ID
//                     command.ExecuteNonQuery();
//                 }
//             }
//         }
//
//         public void Update(Participant participant, int id)
//         {
//             using (var connection = AppDbContext.GetConnection())
//             {
//                 using (var command = new SQLiteCommand("UPDATE Participant SET Name = @Name, EngineCapacity = @EngineCapacity, Team = @Team WHERE ID = @ID", connection))
//                 {
//                     command.Parameters.AddWithValue("@Name", participant.Name);
//                     command.Parameters.AddWithValue("@EngineCapacity", participant.EngineCapacity);
//                     command.Parameters.AddWithValue("@Team", participant.Team);
//                     command.Parameters.AddWithValue("@ID", id);
//                     command.ExecuteNonQuery();
//                 }
//             }
//         }
//
//         public Participant FindById(int id)
//         {
//             using (var connection = AppDbContext.GetConnection())
//             {
//                 using (var command = new SQLiteCommand("SELECT ID, Name, EngineCapacity, Team FROM Participant WHERE ID = @ID", connection))
//                 {
//                     command.Parameters.AddWithValue("@ID", id);
//                     using (var reader = command.ExecuteReader())
//                     {
//                         if (reader.Read())
//                         {
//                             return new Participant(
//                                 reader.GetInt32(0),
//                                 reader.GetString(1),
//                                 reader.GetInt32(2),
//                                 reader.GetString(3)
//                             );
//                         }
//                         throw new KeyNotFoundException($"No Participant found with ID: {id}");
//                     }
//                 }
//             }
//         }
//
//         public IEnumerable<Participant> FindAll()
//         {
//             var participants = new List<Participant>();
//             using (var connection = AppDbContext.GetConnection())
//             {
//                 using (var command = new SQLiteCommand("SELECT ID, Name, EngineCapacity, Team FROM Participant", connection))
//                 using (var reader = command.ExecuteReader())
//                 {
//                     while (reader.Read())
//                     {
//                         participants.Add(new Participant(
//                             reader.GetInt32(0),
//                             reader.GetString(1),
//                             reader.GetInt32(2),
//                             reader.GetString(3)
//                         ));
//                     }
//                 }
//             }
//             return participants;
//         }
//
//         // Implementation of GetAll method as per the IRepository interface
//         public ICollection<Participant> GetAll()
//         {
//             var participants = new List<Participant>();
//             using (var connection = AppDbContext.GetConnection())
//             {
//                 using (var command = new SQLiteCommand("SELECT ID, Name, EngineCapacity, Team FROM Participant", connection))
//                 using (var reader = command.ExecuteReader())
//                 {
//                     while (reader.Read())
//                     {
//                         participants.Add(new Participant(
//                             reader.GetInt32(0),
//                             reader.GetString(1),
//                             reader.GetInt32(2),
//                             reader.GetString(3)
//                         ));
//                     }
//                 }
//             }
//             return participants;
//         }
//     }
// }
/*
using System.Data.SQLite;
using motorcycleApp.DataBase;
using motorcycleApp.Models;
using Serilog;
*/

/*namespace motorcycleApp.Repositories;

public class ParticipantRepository : IRepository<Participant, int> {
    public void Add(Participant entity) {
        string sql = "INSERT INTO Participant (Name, EngineCapacity) VALUES (@Name, @EngineCapacity)";
        try {
            using (var conn = ConnectionManager.GetConnection()) {
                
                Log.Information("Connection established " + conn.Database);

                using (var command = new SQLiteCommand(sql, conn)) {
                    command.Parameters.AddWithValue("@Name", entity.Name);
                    command.Parameters.AddWithValue("@EngineCapacity", entity.EngineCapacity);
                    
                    command.ExecuteNonQuery();
                    Log.Information("Added participant " + entity.Name + " to Database");
                }
            }
        }
        catch (Exception ex) {
            Log.Error(ex, "Error while adding participant " + entity.Name + " to Database");
        }
        
    }

    public void Update(Participant entity) {
        string sql = "UPDATE Participant SET Name=@Name, EngineCapacity=@EngineCapacity WHERE Pid=@Id";

        using (var conn = ConnectionManager.GetConnection()) {
            
            Log.Information("Connection established " + conn.Database);
            using (var command = new SQLiteCommand(sql, conn)) {
                command.Parameters.AddWithValue("@Id", entity.ID);
                command.Parameters.AddWithValue("@Name", entity.Name);
                command.Parameters.AddWithValue("@EngineCapacity", entity.EngineCapacity);
                
                command.ExecuteNonQuery();
            }
        }
        Log.Information("Updated participant " + entity.Name + " to Database");
        
    }

    public void Delete(Participant entity) {
        string sql = "DELETE FROM Participant WHERE Pid=@Id";

        using (var conn = ConnectionManager.GetConnection()) {
            
            Log.Information("Connection established " + conn.Database);

            using (var command = new SQLiteCommand(sql, conn)) {
                command.Parameters.AddWithValue("@Id", entity.ID);
                
                int affectedRows = command.ExecuteNonQuery();
                if (affectedRows > 0) {
                    Log.Information("Deleted participant " + entity.Name + " to Database");
                }
                else {
                    Log.Information("Deleted participant " + entity.Name + " to Database");
                }
            }
        }
    }

    public Participant FindById(int id) {
        string sql = "SELECT * FROM Participant WHERE Pid=@Id";

        using (var conn = ConnectionManager.GetConnection()) {
            
            Log.Information("Connection established " + conn.Database);
            using (var command = new SQLiteCommand(sql, conn)) {
                command.Parameters.AddWithValue("@Id", id);
                using (var reader = command.ExecuteReader()) {
                    if (reader.Read()) {
                        Participant participant = new Participant(
                            reader.GetInt32(0),//id
                            reader.GetString(1),//name
                            reader.GetInt32(2));//engine_cap
                        Log.Information("Found participant " + participant.Name + " to Database");
                        return participant;
                    }
                }
            }
        }
        Log.Warning("Could not find participant " + id + " to Database");
        return null;
    }

    public IEnumerable<Participant> GetAll() {
        string sql = "SELECT * FROM Participant";
        var participants = new List<Participant>();
        
        using (var conn = ConnectionManager.GetConnection()) {
            Log.Information("Connection established " + conn.Database);
            
            using (var command = new SQLiteCommand(sql, conn))
            using (var reader = command.ExecuteReader()) {
                while (reader.Read()) {
                    participants.Add(new Participant(
                        reader.GetInt32(0),//id
                        reader.GetString(1),//name
                        reader.GetInt32(2)
                    ));
                }
            }
        }
        Log.Information("Found " + participants.Count + " participants to Database");
        return participants;
    }
}*/

File: project-java-cosminpac04/motorcycleContest/src/main/java/motorcycle/repository/ParticipantRepository.java
package motorcycle.repository;

import motorcycle.database.DataBaseConnection;
import motorcycle.model.Participant;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.sql.*;
import java.util.ArrayList;
import java.util.List;

public class ParticipantRepository implements IRepository<Participant, Integer> {
    private static final Logger logger = LoggerFactory.getLogger(ParticipantRepository.class);

    @Override
    public void add(Participant participant) {
        String sql = "INSERT INTO Participant (Name, EngineCapacity, Team) VALUES (?, ?, ?)";
        try (Connection conn = DataBaseConnection.connect();
             PreparedStatement stmt = conn.prepareStatement(sql)) {

            stmt.setString(1, participant.getName());
            stmt.setInt(2, participant.getEngineCapacity());
            stmt.setString(3, participant.getTeam());
            stmt.executeUpdate();

            logger.info("Added participant: {}", participant.getName());

        } catch (SQLException e) {
            logger.error("Error adding participant: {}", participant.getName(), e);
        }
    }

    @Override
    public void delete(Participant participant) {
        String sql = "DELETE FROM Participant WHERE Pid = ?";
        try (Connection conn = DataBaseConnection.connect();
             PreparedStatement stmt = conn.prepareStatement(sql)) {

            stmt.setInt(1, participant.getID());
            int rowsAffected = stmt.executeUpdate();

            if (rowsAffected > 0) {
                logger.info("Deleted participant with ID: {}", participant.getID());
            } else {
                logger.warn("No participant found with ID: {}", participant.getID());
            }

        } catch (SQLException e) {
            logger.error("Error deleting participant with ID: {}", participant.getID(), e);
        }
    }

    @Override
    public void update(Participant participant, Integer id) {
        String sql = "UPDATE Participant SET Name = ?, EngineCapacity = ?, Team = ? WHERE Pid = ?";
        try (Connection conn = DataBaseConnection.connect();
             PreparedStatement stmt = conn.prepareStatement(sql)) {

            stmt.setString(1, participant.getName());
            stmt.setInt(2, participant.getEngineCapacity());
            stmt.setString(3, participant.getTeam());
            stmt.setInt(4, id);
            stmt.executeUpdate();

            logger.info("Updated participant with ID: {}", id);

        } catch (SQLException e) {
            logger.error("Error updating participant with ID: {}", id, e);
        }
    }

    @Override
    public Participant findById(Integer Pid) {
        String sql = "SELECT * FROM Participant WHERE Pid = ?";
        try (Connection conn = DataBaseConnection.connect();
             PreparedStatement stmt = conn.prepareStatement(sql)) {

            stmt.setInt(1, Pid);
            ResultSet rs = stmt.executeQuery();

            if (rs.next()) {
                Participant participant = new Participant(
                        rs.getInt("Pid"),
                        rs.getString("name"),
                        rs.getInt("EngineCapacity"),
                        rs.getString("team")
                );
                logger.info("Found participant: {}", participant.getName());
                return participant;
            } else {
                logger.warn("No participant found with ID: {}", Pid);
            }

        } catch (SQLException e) {
            logger.error("Error finding participant with ID: {}", Pid, e);
        }
        return null;
    }

    @Override
    public List<Participant> findAll() {
        List<Participant> participants = new ArrayList<>();
        String sql = "SELECT * FROM Participant";

        try (Connection conn = DataBaseConnection.connect();
             Statement stmt = conn.createStatement();
             ResultSet rs = stmt.executeQuery(sql)) {

            while (rs.next()) {
                participants.add(new Participant(
                        rs.getInt("Pid"),
                        rs.getString("Name"),
                        rs.getInt("EngineCapacity"),
                        rs.getString("Team")
                ));
            }

            logger.info("Retrieved {} participants", participants.size());

        } catch (SQLException e) {
            logger.error("Error retrieving participants", e);
        }
        return participants;
    }

    @Override
    public List<Participant> getAll() {
        return findAll();
    }
}
